File: src-tauri/src/reddit/parser.rs
use serde_json::Value;

use super::types::{FetchResult, MediaItem, MediaPost, MediaType};

pub fn parse_listing(listing: &Value) -> FetchResult {
    let after = listing["data"]["after"].as_str().map(String::from);

    let posts: Vec<MediaPost> = listing["data"]["children"]
        .as_array()
        .map(|children| {
            children
                .iter()
                .filter_map(|child| parse_post(&child["data"]))
                .collect()
        })
        .unwrap_or_default();

    FetchResult { posts, after }
}

fn parse_post(post: &Value) -> Option<MediaPost> {
    // Skip self/text posts
    if post["is_self"].as_bool().unwrap_or(false) {
        return None;
    }

    let id = post["id"].as_str()?.to_string();
    let title = post["title"].as_str().unwrap_or("").to_string();
    let author = post["author"].as_str().unwrap_or("[deleted]").to_string();
    let score = post["score"].as_i64().unwrap_or(0);
    let num_comments = post["num_comments"].as_u64().unwrap_or(0);
    let permalink = post["permalink"].as_str().unwrap_or("").to_string();
    let subreddit = post["subreddit"].as_str().unwrap_or("").to_string();
    let over_18 = post["over_18"].as_bool().unwrap_or(false);
    let url = post["url"].as_str().unwrap_or("");
    let domain = post["domain"].as_str().unwrap_or("");
    let post_hint = post["post_hint"].as_str().unwrap_or("");

    let base = MediaPost {
        id,
        title,
        author,
        score,
        num_comments,
        permalink,
        subreddit,
        over_18,
        media_type: MediaType::Image,
        media: vec![],
        audio_url: None,
        embed_url: None,
    };

    // 1. Gallery
    if post["is_gallery"].as_bool().unwrap_or(false) {
        if let Some(result) = try_gallery(post, base.clone()) {
            return Some(result);
        }
    }

    // 2. Reddit video (v.redd.it)
    if post["is_video"].as_bool().unwrap_or(false) {
        if let Some(result) = try_reddit_video(post, base.clone()) {
            return Some(result);
        }
    }

    // 3. Direct image URL
    let url_lower = url.to_lowercase();
    if url_lower.ends_with(".jpg")
        || url_lower.ends_with(".jpeg")
        || url_lower.ends_with(".png")
        || url_lower.ends_with(".webp")
    {
        return Some(MediaPost {
            media_type: MediaType::Image,
            media: vec![MediaItem {
                url: url.to_string(),
                width: None,
                height: None,
                caption: None,
            }],
            ..base
        });
    }

    // 4. Direct gif URL
    if url_lower.ends_with(".gif") {
        return Some(MediaPost {
            media_type: MediaType::Image,
            media: vec![MediaItem {
                url: url.to_string(),
                width: None,
                height: None,
                caption: None,
            }],
            ..base
        });
    }

    // 5. i.redd.it or i.imgur.com domain
    if domain == "i.redd.it" || domain == "i.imgur.com" {
        return Some(MediaPost {
            media_type: MediaType::Image,
            media: vec![MediaItem {
                url: url.to_string(),
                width: None,
                height: None,
                caption: None,
            }],
            ..base
        });
    }

    // 6. imgur.com (no extension, not album) -> transform to direct image
    if domain == "imgur.com" && !url.contains("/a/") && !url.contains("/gallery/") {
        let imgur_id = url.rsplit('/').next().unwrap_or("");
        if !imgur_id.is_empty() {
            return Some(MediaPost {
                media_type: MediaType::Image,
                media: vec![MediaItem {
                    url: format!("https://i.imgur.com/{}.jpg", imgur_id),
                    width: None,
                    height: None,
                    caption: None,
                }],
                ..base
            });
        }
    }

    // 7. .gifv -> .mp4
    if url_lower.ends_with(".gifv") {
        let mp4_url = format!("{}.mp4", &url[..url.len() - 5]);
        return Some(MediaPost {
            media_type: MediaType::AnimatedGif,
            media: vec![MediaItem {
                url: mp4_url,
                width: None,
                height: None,
                caption: None,
            }],
            ..base
        });
    }

    // 8. Direct .mp4
    if url_lower.ends_with(".mp4") {
        return Some(MediaPost {
            media_type: MediaType::Video,
            media: vec![MediaItem {
                url: url.to_string(),
                width: None,
                height: None,
                caption: None,
            }],
            ..base
        });
    }

    // 9. post_hint == "image" with preview
    if post_hint == "image" {
        if let Some(preview_url) = post["preview"]["images"][0]["source"]["url"].as_str() {
            return Some(MediaPost {
                media_type: MediaType::Image,
                media: vec![MediaItem {
                    url: preview_url.to_string(),
                    width: post["preview"]["images"][0]["source"]["width"].as_u64().map(|v| v as u32),
                    height: post["preview"]["images"][0]["source"]["height"].as_u64().map(|v| v as u32),
                    caption: None,
                }],
                ..base
            });
        }
    }

    // 10. YouTube
    if domain.contains("youtube.com") || domain.contains("youtu.be") {
        if let Some(embed) = youtube_embed_url(url) {
            return Some(MediaPost {
                media_type: MediaType::Embed,
                media: vec![],
                embed_url: Some(embed),
                ..base
            });
        }
    }

    // 11. Redgifs
    if domain.contains("redgifs.com") {
        if let Some(embed) = redgifs_embed_url(url) {
            return Some(MediaPost {
                media_type: MediaType::Embed,
                media: vec![],
                embed_url: Some(embed),
                ..base
            });
        }
    }

    // 12. rich:video with secure_media_embed
    if post_hint == "rich:video" {
        if let Some(content) = post["secure_media_embed"]["content"].as_str() {
            if let Some(src) = extract_iframe_src(content) {
                return Some(MediaPost {
                    media_type: MediaType::Embed,
                    media: vec![],
                    embed_url: Some(src),
                    ..base
                });
            }
        }
    }

    // 13. Skip everything else
    None
}

fn try_gallery(post: &Value, base: MediaPost) -> Option<MediaPost> {
    let metadata = post.get("media_metadata")?;
    let gallery_items = post["gallery_data"]["items"].as_array()?;

    let mut media = Vec::new();

    for item in gallery_items {
        let media_id = item["media_id"].as_str()?;
        let entry = metadata.get(media_id)?;

        // Check status
        if entry["status"].as_str() != Some("valid") {
            continue;
        }

        // Get the best quality image: s (source) field
        let source = &entry["s"];
        let url = source["u"]
            .as_str()
            .or_else(|| source["gif"].as_str())
            .or_else(|| source["mp4"].as_str())?;

        let width = source["x"].as_u64().map(|v| v as u32);
        let height = source["y"].as_u64().map(|v| v as u32);
        let caption = item["caption"].as_str().map(String::from);

        media.push(MediaItem {
            url: url.to_string(),
            width,
            height,
            caption,
        });
    }

    if media.is_empty() {
        return None;
    }

    Some(MediaPost {
        media_type: MediaType::Gallery,
        media,
        ..base
    })
}

fn try_reddit_video(post: &Value, base: MediaPost) -> Option<MediaPost> {
    let reddit_video = post["media"]["reddit_video"].as_object()?;
    let fallback_url = reddit_video.get("fallback_url")?.as_str()?;
    let width = reddit_video
        .get("width")
        .and_then(|v| v.as_u64())
        .map(|v| v as u32);
    let height = reddit_video
        .get("height")
        .and_then(|v| v.as_u64())
        .map(|v| v as u32);
    let is_gif = reddit_video
        .get("is_gif")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);

    // Construct audio URL - try both naming conventions
    let audio_url = if !is_gif {
        // Extract base URL before DASH_
        if let Some(dash_pos) = fallback_url.rfind("DASH_") {
            let base_url = &fallback_url[..dash_pos];
            Some(format!("{}DASH_AUDIO_128.mp4", base_url))
        } else {
            None
        }
    } else {
        None
    };

    Some(MediaPost {
        media_type: if is_gif {
            MediaType::AnimatedGif
        } else {
            MediaType::Video
        },
        media: vec![MediaItem {
            url: fallback_url.to_string(),
            width,
            height,
            caption: None,
        }],
        audio_url,
        ..base
    })
}

fn youtube_embed_url(url: &str) -> Option<String> {
    // youtube.com/watch?v=ID
    if let Some(pos) = url.find("v=") {
        let id = &url[pos + 2..];
        let id = id.split('&').next().unwrap_or(id);
        return Some(format!("https://www.youtube.com/embed/{}", id));
    }
    // youtu.be/ID
    if url.contains("youtu.be/") {
        let id = url.rsplit("youtu.be/").next()?;
        let id = id.split('?').next().unwrap_or(id);
        return Some(format!("https://www.youtube.com/embed/{}", id));
    }
    None
}

fn redgifs_embed_url(url: &str) -> Option<String> {
    // redgifs.com/watch/slug -> redgifs.com/ifr/slug
    if let Some(slug) = url.rsplit("/watch/").next() {
        let slug = slug.split('?').next().unwrap_or(slug);
        return Some(format!("https://www.redgifs.com/ifr/{}", slug));
    }
    None
}

fn extract_iframe_src(html: &str) -> Option<String> {
    // Simple extraction of src="..." from iframe HTML
    let src_pos = html.find("src=\"").or_else(|| html.find("src='"))?;
    let quote_char = html.as_bytes()[src_pos + 4] as char;
    let start = src_pos + 5;
    let rest = &html[start..];
    let end = rest.find(quote_char)?;
    Some(rest[..end].to_string())
}

File: src/reddit/parser.ts
import type { FetchResult, MediaItem, MediaPost } from "../types";

export function parseListing(listing: any): FetchResult {
  const after: string | null = listing?.data?.after ?? null;

  const children: any[] = listing?.data?.children ?? [];
  const posts: MediaPost[] = children
    .map((child: any) => parsePost(child?.data))
    .filter((p: MediaPost | null): p is MediaPost => p !== null);

  return { posts, after };
}

function parsePost(post: any): MediaPost | null {
  if (!post) return null;
  if (post.is_self) return null;

  const id: string | undefined = post.id;
  if (!id) return null;

  const title: string = post.title ?? "";
  const author: string = post.author ?? "[deleted]";
  const score: number = post.score ?? 0;
  const num_comments: number = post.num_comments ?? 0;
  const permalink: string = post.permalink ?? "";
  const subreddit: string = post.subreddit ?? "";
  const over_18: boolean = post.over_18 ?? false;
  const url: string = post.url ?? "";
  const domain: string = post.domain ?? "";
  const postHint: string = post.post_hint ?? "";

  const base: MediaPost = {
    id, title, author, score, num_comments, permalink, subreddit, over_18,
    media_type: "image", media: [], audio_url: null, embed_url: null,
  };

  // 1. Gallery
  if (post.is_gallery) {
    const result = tryGallery(post, base);
    if (result) return result;
  }

  // 2. Reddit video (v.redd.it)
  if (post.is_video) {
    const result = tryRedditVideo(post, base);
    if (result) return result;
  }

  // 3. Direct image URL
  const urlLower = url.toLowerCase();
  if (urlLower.endsWith(".jpg") || urlLower.endsWith(".jpeg") ||
      urlLower.endsWith(".png") || urlLower.endsWith(".webp")) {
    return { ...base, media: [{ url, width: null, height: null, caption: null }] };
  }

  // 4. Direct gif URL
  if (urlLower.endsWith(".gif")) {
    return { ...base, media: [{ url, width: null, height: null, caption: null }] };
  }

  // 5. i.redd.it or i.imgur.com domain
  if (domain === "i.redd.it" || domain === "i.imgur.com") {
    return { ...base, media: [{ url, width: null, height: null, caption: null }] };
  }

  // 6. imgur.com (no extension, not album)
  if (domain === "imgur.com" && !url.includes("/a/") && !url.includes("/gallery/")) {
    const imgurId = url.split("/").pop() ?? "";
    if (imgurId) {
      return { ...base, media: [{ url: `https://i.imgur.com/${imgurId}.jpg`, width: null, height: null, caption: null }] };
    }
  }

  // 7. .gifv -> .mp4
  if (urlLower.endsWith(".gifv")) {
    return { ...base, media_type: "animated_gif", media: [{ url: url.slice(0, -5) + ".mp4", width: null, height: null, caption: null }] };
  }

  // 8. Direct .mp4
  if (urlLower.endsWith(".mp4")) {
    return { ...base, media_type: "video", media: [{ url, width: null, height: null, caption: null }] };
  }

  // 9. post_hint == "image" with preview
  if (postHint === "image") {
    const previewUrl: string | undefined = post.preview?.images?.[0]?.source?.url;
    if (previewUrl) {
      return { ...base, media: [{
        url: previewUrl,
        width: post.preview?.images?.[0]?.source?.width ?? null,
        height: post.preview?.images?.[0]?.source?.height ?? null,
        caption: null,
      }] };
    }
  }

  // 10. YouTube
  if (domain.includes("youtube.com") || domain.includes("youtu.be")) {
    const embed = youtubeEmbedUrl(url);
    if (embed) return { ...base, media_type: "embed", embed_url: embed };
  }

  // 11. Redgifs — store slug for async resolution
  if (domain.includes("redgifs.com")) {
    const slug = redgifsSlug(url);
    if (slug) return { ...base, media_type: "embed", embed_url: `redgifs:${slug}` };
  }

  // 12. rich:video with secure_media_embed
  if (postHint === "rich:video") {
    const content: string | undefined = post.secure_media_embed?.content;
    if (content) {
      const src = extractIframeSrc(content);
      if (src) return { ...base, media_type: "embed", embed_url: src };
    }
  }

  return null;
}

function tryGallery(post: any, base: MediaPost): MediaPost | null {
  const metadata = post.media_metadata;
  const galleryItems: any[] | undefined = post.gallery_data?.items;
  if (!metadata || !galleryItems) return null;

  const media: MediaItem[] = [];
  for (const item of galleryItems) {
    const mediaId: string | undefined = item.media_id;
    if (!mediaId) return null;
    const entry = metadata[mediaId];
    if (!entry) return null;
    if (entry.status !== "valid") continue;
    const source = entry.s;
    const url: string | undefined = source?.u ?? source?.gif ?? source?.mp4;
    if (!url) return null;
    media.push({ url, width: source?.x ?? null, height: source?.y ?? null, caption: item.caption ?? null });
  }
  if (media.length === 0) return null;
  return { ...base, media_type: "gallery", media };
}

function tryRedditVideo(post: any, base: MediaPost): MediaPost | null {
  const redditVideo = post.media?.reddit_video;
  if (!redditVideo) return null;
  const fallbackUrl: string | undefined = redditVideo.fallback_url;
  if (!fallbackUrl) return null;
  const isGif: boolean = redditVideo.is_gif ?? false;
  let audioUrl: string | null = null;
  if (!isGif) {
    const dashPos = fallbackUrl.lastIndexOf("DASH_");
    if (dashPos !== -1) audioUrl = fallbackUrl.slice(0, dashPos) + "DASH_AUDIO_128.mp4";
  }
  return {
    ...base,
    media_type: isGif ? "animated_gif" : "video",
    media: [{ url: fallbackUrl, width: redditVideo.width ?? null, height: redditVideo.height ?? null, caption: null }],
    audio_url: audioUrl,
  };
}

function youtubeEmbedUrl(url: string): string | null {
  const vPos = url.indexOf("v=");
  if (vPos !== -1) return `https://www.youtube.com/embed/${url.slice(vPos + 2).split("&")[0]}`;
  if (url.includes("youtu.be/")) {
    const id = (url.split("youtu.be/").pop() ?? "").split("?")[0];
    if (id) return `https://www.youtube.com/embed/${id}`;
  }
  return null;
}

function redgifsSlug(url: string): string | null {
  const parts = url.split("/watch/");
  if (parts.length < 2) return null;
  const slug = parts[parts.length - 1].split("?")[0];
  return slug ? slug.toLowerCase() : null;
}

function extractIframeSrc(html: string): string | null {
  const pos = html.indexOf('src="') !== -1 ? html.indexOf('src="') : html.indexOf("src='");
  if (pos === -1) return null;
  const quoteChar = html[pos + 4];
  const rest = html.slice(pos + 5);
  const end = rest.indexOf(quoteChar);
  return end === -1 ? null : rest.slice(0, end);
}
